src/index.ts

```diff
@@ -16,213 +16,91 @@
-  // 3. Environment Variable Set Configuration File (from GCLOUD_TRACE_CONFIG)
-  // 4. Default Config (as specified in './config')
-  const config = extend(
-      true, {[FORCE_NEW]: projectConfig[FORCE_NEW]}, defaultConfig,
-      envSetConfig, projectConfig, envConfig, {plugins: {}});
-  // The empty plugins object guarantees that plugins is a plain object,
-  // even if it's explicitly specified in the config to be a non-object.
-
-  // Enforce the upper limit for the label value size.
-  if (config.maximumLabelValueSize >
-      Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT) {
-    config.maximumLabelValueSize = Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT;
-  }
-  // Clamp the logger level.
-  if (config.logLevel < 0) {
-    config.logLevel = 0;
-  } else if (config.logLevel >= common.logger.LEVELS.length) {
-    config.logLevel = common.logger.LEVELS.length - 1;
-  }
-  return config;
-}
-
-/**
- * Stops the Trace Agent. This disables the publicly exposed agent instance,
- * as well as any instances passed to plugins. This also prevents the Trace
- * Writer from publishing additional traces.
- */
-function stop() {
-  if (pluginLoader.exists()) {
-    pluginLoader.get().deactivate();
-  }
-  if (traceAgent && traceAgent.isActive()) {
-    traceAgent.disable();
-  }
-  if (cls.exists()) {
-    cls.get().disable();
-  }
-  if (traceWriter.exists()) {
-    traceWriter.get().stop();
-  }
-}
-
-/**
- * Start the Trace agent that will make your application available for
- * tracing with Stackdriver Trace.
- *
- * @param config - Trace configuration
+ * Start the Stackdriver Trace Agent with the given configuration (if provided).
+ * This function should only be called once, and before any other modules are
+ * loaded.
+ * @param config A configuration object.
+ * @returns An object exposing functions for creating custom spans.
  *
  * @resource [Introductory video]{@link
  * https://www.youtube.com/watch?v=NCFDqeo7AeY}
  *
  * @example
  * trace.start();
  */
-export function start(projectConfig?: Config): PluginTypes.TraceAgent {
-  const config = initConfig(projectConfig || {});
-
-  if (traceAgent.isActive() && !config[FORCE_NEW]) {  // already started.
-    throw new Error('Cannot call start on an already started agent.');
-  } else if (traceAgent.isActive()) {
-    // For unit tests only.
-    // Undoes initialization that occurred last time start() was called.
-    stop();
-  }
-
-  if (!config.enabled) {
-    return traceAgent;
+export function start(config?: Config): PluginTypes.TraceAgent {
+  // Determine the preferred context propagation mechanism, as
+  // continuation-local-storage should be loaded before any modules that do I/O.
+  const ahAvailable = semver.satisfies(process.version, '>=8');
+  const agentEnabled = !config || config.enabled !== false;
+  const alAutoPreferred =
+      !ahAvailable && (!config || config.clsMechanism === 'auto');
+  const alUserPreferred = config && (config.clsMechanism === 'async-listener');
```

Comment:
Why not:
1. Keep `initConfig` in this file. As far as I can tell, it is synchronous and doesn't need external modules.
2. Modify `initConfig` with additional logic to convert the user preferences into a concrete (non-`auto`) value of `config.mechanism`.

In other words, I would prefer if `start` had logic like this:

```ts
function start(config?: Config) {
  const normalizedConfig = initConfig(config);
  if (!config.enabled) return;
  if (config.mechanism == 'async-listener') require('continuation-local-storage');
  // ...
}
```

The way you have written it right now, we don't properly deal with a config file provided through an environment variable.

Reply:
You may need to change the `require` in `initConfig` to a `fs.readFileSync`.

If the above proposal is acceptable, are the rest of the changes (refactoring most of the logic in this file into `tracing.ts`) still needed? This is a pretty large change, and careful reasoning is needed to make sure it doesn't break behaviour. It is not immediately obvious to me why the refactoring is *required*.

Reply:
The risk with the `require` that I was thinking of was if the users put arbitrary JS in there which ends up loading modules. I am not sure if it an important enough concern.

> For the rest of the refactoring -- I had been thinking about whether class-ifying Tracing was going overboard -- it's not strictly necessary

If it is not strictly necessary, I would prefer not to see it in *this* PR. This PR should look obviously correct to me, and currently it would require a fairly in depth review to assert that.

Reply:
To clarify, I believe moving logic to a separate file is the necessary part, while making it a class is less necessary. That being said, I think the change to make it a class is not going to negatively impact the diff view by a lot.